src/java.base/share/classes/com/sun/crypto/provider/PBMAC1Parameters.java

```diff
@@ -0,0 +1,254 @@
+
+    // the salt
+    private byte[] salt = null;
+
+    // the iteration count
+    private int iCount = 0;
+
+    // the Hmac function used by the kdf
+    private String kdfHmac = null;
+
+    // the Hmac function
+    private String Hmac = null;
+
+    // the key derivation function (default is HmacSHA1)
+    private final ObjectIdentifier kdfAlgo_OID =
+            ObjectIdentifier.of(KnownOIDs.HmacSHA1);
+
+    // length of key generated by the kdf
+    private int keyLength = -1;
+
+    protected void engineInit(AlgorithmParameterSpec paramSpec)
+        throws InvalidParameterSpecException
+    {
+       if (!(paramSpec instanceof PBMAC1ParameterSpec)) {
+           throw new InvalidParameterSpecException
+               ("Inappropriate parameter specification");
+       }
+       salt = ((PBMAC1ParameterSpec)paramSpec).getSalt().clone();
+       iCount = ((PBMAC1ParameterSpec)paramSpec).getIterationCount();
+    }
+
+    protected void engineInit(byte[] encoded)
+        throws IOException
+    {
+        DerValue pBMAC1_params = new DerValue(encoded);
+        if (pBMAC1_params.tag != DerValue.tag_Sequence) {
+            throw new IOException("PMAC1 parameter parsing error: "
+                + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue[] Info = (new DerInputStream(pBMAC1_params.toByteArray()))
+                .getSequence(2);
+        if (Info.length != 2) {
+            throw new IOException("PMAC1 parameter parsing error: "
+                + "expected length not 2");
+        }
+        ObjectIdentifier OID = Info[1].data.getOID();
+            KnownOIDs o = KnownOIDs.findMatch(OID.toString());
+            if (o == null || (!o.stdName().equals("HmacSHA1") &&
+                    !o.stdName().equals("HmacSHA224") &&
+                    !o.stdName().equals("HmacSHA256") &&
+                    !o.stdName().equals("HmacSHA384") &&
+                    !o.stdName().equals("HmacSHA512") &&
+                    !o.stdName().equals("HmacSHA512/224") &&
+                    !o.stdName().equals("HmacSHA512/256"))) {
+                throw new IOException("PBMAC1 parameter parsing error: "
+                        + "expecting the object identifier for a HmacSHA key "
+                        + "derivation function");
+            }
+            Hmac = o.stdName();
+
+        DerValue kdf = pBMAC1_params.data.getDerValue();
+        var kdfParams = new PBKDF2Parameters();
+        String kdfAlgo = kdfParams.parseKDF(kdf);
+        salt = kdfParams.getSalt();
+        iCount = kdfParams.getIterationCount();
+
+        // Key length SHOULD be the same size as the HMAC function output size.
+        keyLength = kdfParams.getKeyLength();
+
+        // Implementations MUST NOT accept PBKDF2 KDF with params that
+        // omit the keyLength field.
+        if (keyLength == -1) {
+            throw new IOException("PMAC1 parameter parsing error: "
+                + "missing keyLength field");
+        }
```

Comment:
Where is these requirements on `keyLength` documented?

Reply:
You are right. The check belongs in the `PKCS12KeyStore` class.

Reply:
Since I've replaced usage of the `PBMAC1ParameterSpec` with `PBEParameterSpec`, there is no longer a way to pass keyLength to `PKCS12KeyStore`. I think the HMAC output size check can be removed since it is a SHOULD. The keyLength is present check should stay for now since it is a MUST. We can revisit this later if we switch to a different parameter spec.